classes/controller/AdminController.php

```diff
@@ -4789,4 +4795,80 @@ protected function isAnonymousAllowed()
     {
         return $this->allowAnonymous;
     }
+
+    /**
+     * Get the buttons provided by hooks and merge them into the already defined page toolbar buttons
+     *
+     * @throws TypeException
+     */
+    protected function mergeExtraToolbarButtons(): void
+    {
+        $toolbarButtonsCollection = new ActionsBarButtonsCollection();
+
+        // Get previously assigned toolbar buttons of the controller
+        $controllerButtons = $this->page_header_toolbar_btn;
+        if (!empty($controllerButtons)) {
+            // Build ActionsBarButton based on array setting from the controller and add it to collection
+            foreach ($controllerButtons as $controllerButton) {
+                $toolbarButtonsCollection->add(
+                    new ActionsBarButton(
+                        $controllerButton['class'] ?? '',
+                        $controllerButton,
+                        $controllerButton['desc'] ?? ''
+                    )
+                );
+            }
+        }
+
+        // Get the toolbar buttons defined by hooks and add them to the main collection
+        try {
+            Hook::exec('actionGetAdminToolbarButtons', [
+                'controller' => $this,
+                'toolbar_extra_buttons_collection' => &$toolbarButtonsCollection,
+            ]);
+        } catch (Exception $exception) {
+            $this->get('logger')->addWarning(
+                'There was an error retrieving toolbar buttons from Hooks. Toolbar buttons are probably not complete',
+                [
+                    'message' => $exception->getMessage(),
+                    'file' => $exception->getFile(),
+                    'line' => $exception->getLine(),
+                ]
+            );
+
+            return; // In case of any error, the buttons from hooks are simply ignored
+        }
+
+        // Transform the collection of buttons into array and reassign it to the page header toolbar buttons
+        $toolbarButtons = [];
+
+        foreach ($toolbarButtonsCollection as $toolbarActionButton) {
+            $toolbarButtons = array_merge(
+                $toolbarButtons,
+                $this->transformActionBarButtonToToolbarButton($toolbarActionButton)
+            );
+        }
```

Comment:
Just a suggestion but isn't it simpler to do:
```suggestion
            $toolbarButtons[$toolbarActionButton->getClass()] = $this->transformActionBarButtonToToolbarButton($toolbarActionButton);
```

Reply:
Also the fact that we use the class as a key means the buttons could be erased when they have the same class We should have a dedicated `getId()` method to tell them apart